pkg/server/handler/auth.go

```diff
@@ -1058,50 +1058,118 @@ func (h *LinkProviderHandler) Handle(payload *router.Payload, response *router.R
-		panic(err)
+func (payload *unlinkProviderPayload) Validate() skyerr.Error {
+	if payload.Provider == "" {
+		return skyerr.NewInvalidArgument("empty provider", []string{"provider"})
 	}
 
-	if err = store.Put(&token); err != nil {
-		panic(err)
+	if payload.UserID == "" {
+		return skyerr.NewInvalidArgument("empty user id", []string{"user_id"})
 	}
 
-	authResponse, err := AuthResponseFactory{
-		AssetStore: h.AssetStore,
-		Conn:       payload.DBConn,
-	}.NewAuthResponse(info, user, token.AccessToken, payload.HasMasterKey())
-	if err != nil {
-		response.Err = skyerr.MakeError(err)
+	return nil
+}
+
+// UnlinkProviderHandler disconnect user with specific provider
+//
+// UnlinkProviderHandler receives parameters:
+//
+// * provider (string, required)
+// * user_id (string, required)
+//
+// curl -X POST -H "Content-Type: application/json" \
+//   -d @- http://localhost:3000/ <<EOF
+// {
+// 		"action": "auth:provider:unlink",
+// 		"provider": "facebook",
+// 		"user_id": "c0959b6b-15ea-4e21-8afb-9c8308ad79db"
+// }
+// EOF
+// Response
+// {
+//     "result": "OK"
+// }
+type UnlinkProviderHandler struct {
+	HookRegistry     *hook.Registry     `inject:"HookRegistry"`
+	AssetStore       asset.Store        `inject:"AssetStore"`
+	AuthRecordKeys   [][]string         `inject:"AuthRecordKeys"`
+	AccessKey        router.Processor   `preprocessor:"accesskey"`
+	DBConn           router.Processor   `preprocessor:"dbconn"`
+	InjectPublicDB   router.Processor   `preprocessor:"inject_public_db"`
+	PluginReady      router.Processor   `preprocessor:"plugin_ready"`
+	RequireMasterKey router.Processor   `preprocessor:"require_master_key"`
+	preprocessors    []router.Processor
+}
+
+func (h *UnlinkProviderHandler) Setup() {
+	h.preprocessors = []router.Processor{
+		h.AccessKey,
+		h.DBConn,
+		h.InjectPublicDB,
+		h.PluginReady,
+		h.RequireMasterKey,
+	}
+}
+
+func (h *UnlinkProviderHandler) GetPreprocessors() []router.Processor {
+	return h.preprocessors
+}
+
+func (h *UnlinkProviderHandler) Handle(payload *router.Payload, response *router.Response) {
+	log.Debugf("Unlink provider")
+	p := &unlinkProviderPayload{}
+	skyErr := p.Decode(payload.Data)
+	if skyErr != nil {
+		response.Err = skyErr
 		return
 	}
 
-	// Populate the activity time to user
-	now := timeNow()
-	info.LastSeenAt = &now
-	if err := payload.DBConn.UpdateAuth(&info); err != nil {
-		response.Err = skyerr.MakeError(err)
+	info := skydb.AuthInfo{}
+	userID := p.UserID
+
+	if err := payload.DBConn.GetAuth(userID, &info); err != nil {
+		response.Err = skyerr.NewError(skyerr.ResourceNotFound, "user not found")
 		return
 	}
 
-	// update user record last login time
-	user.UpdatedAt = now
-	user.UpdaterID = info.ID
-	if p.IsLogin {
-		user.Data[UserRecordLastLoginAtKey] = now
+	providerInfo := info.ProviderInfo
+	unlinked := false
+	for k := range providerInfo {
+		if strings.HasPrefix(k, p.PrincipalIDPrefix) {
+			info.RemoveProviderInfoData(k)
+			unlinked = true
+		}
 	}
```

Comment:
`RemoveProviderInfoData`  should return error for key not found. The iteration is not belong to handler.